008-auth.md

```diff
@@ -0,0 +1,254 @@
+# Authentication and Authorization
+
+## Summary
+
+Right now, the generation of client authentication keys and RBACs thereof is a
+very manual process which is not easy to manipulate for the average user.
+Moreover, it does not encourage best practices such as having protection on
+private keys (via hardware tokens or passwords), shorter-duration certificates,
+user-oriented keys, or even the use of privilege-reduced operational certs.
+
+We should build some tooling to facilitate all of these needs and encourage best
+practices generally.
+
+Design goals:
+- simplify processes
+   - provide easy generation of client keypairs
+     - provide options for protecting client private keys:
+       - password-protection
+       - hardware keys
+   - provide easy generation of user certificates
+     - generate CSR from `talosctl`
+     - generate certificate from CSR using `talosctl`
+     - apply RBAC policies to certificates
+   - provide easy way to generate and update CRLs
+     - figure out how and where to house CRLs
+       - public service? (what if network is not available?)
+       - in etcd? (what if etcd is down?)
+       - on each node? (what if nodes are down during publish?)
+- greater authentication flexibility
+  - support external sources of authentication, such as OIDC servers (dex,
+    Google, etc)
+  - allow key pairs and certificates to be easily bound to managers instead of
+    roles
+- greater authorization flexibility
+  - allow decoupling roles from key sets
+  - allow dynamic and arbitrary mappings of authentication data to roles
+
+Related issues:
+ - https://github.com/talos-systems/talos/issues/3306
+
+## Case Studies
+
+### Any larger organisation
+
+In general, passing authentication credentials around on the network is an
+anti-pattern.
+Once the set of managers of a system grows beyond one, greater flexibility in
+the distribution and management of key and assignment material is necessary.
+
+### Organisations with an existing SSO system
+
+When an organisation already has in place a common authentication system,
+we should be able to externalise the sourcing of our authentication to their
+system.
+However, we still need to be able to control mappings of those identities to
+Talos roles.
```

Comment:
Talos should probably take inspiration from how kube-apiserver can get configured - have `--oidc-groups-claim` and `--oidc-groups-prefix` parameters, and ClusterRole etc. to grant access for certain groups.

I'm not sure where Talos stores its own RBAC stuff, but maybe we can already leverage a lot of this.

Reply:
Talos RBAC is encoded within the client certificate, which would be part of the server-generated artifacts generated upon validation of an external authentication assertion.
Part of the point of this change is to separate the authentication from the authorization... but authorization _must_ be controlled on the server side; client-side authorization is intended to be untrusted.
The authenticator may assert groups, for instance, but it is the role of the Talos server to associate groups from an authenticator to roles within Talos.